### PR TITLE
Implement local cache for crowdsec

### PR DESCRIPTION
**Pull request by maxlerebourg**
https://github.com/fbonalair/traefik-crowdsec-bouncer/pull/32 
With better implementation

3 mode:
- stream mode: use /decisions/stream to get in cache the decisions.
- live mode: like before but with cached response from crowdsec
- none mode: like before (every request ask crowdsec)

### CODE REVIEW

File: controler/controler.go [hunk 1]
```diff
@@ -119,24 +183,52 @@ func ForwardAuth(c *gin.Context) {
...
+		}
 		c.Status(http.StatusOK)
 	}
 }
 
 /*
-	Route to check bouncer connectivity with Crowdsec agent. Mainly use for Kubernetes readiness probe
+Route to check bouncer connectivity with Crowdsec agent. Mainly use for Kubernetes readiness probe
 */
 func Healthz(c *gin.Context) {
```
Comment (fbonalair) [inline controler/controler.go:229]: Adding cache to our service break it's statelessness, it should be reflected in this probe.   
Could you implement something like "In stream mode, while the go routine HandleStreamCache hasn't successfully completed a run, this route return a HTTP 503".

**Review by fbonalair — CHANGES_REQUESTED** [review]
#21 
Thank you a lot for the work !    
Excellent idea for the go routine around the cache.   
I added some quality of life comments around, feel free to check them out.

On more major note though is in the steam mode: make the server fail in case of error. For the bouncer I took a very defensive approach - it's a security component after all - we don't want a bouncer that is allowing any request because we couldn't update the cache so **[fail-fast](https://en.wikipedia.org/wiki/Fail-fast)**. Thanks to you, user have other mode as fallback if they can't found/fix the issue.   
For that in the function HandleStreamCache, you have a lot of returns in case of error but without handling them. In those case Panic with a descriptive log and make the whole bouncer fail.

File: controler/controler.go [hunk 2]
```diff
@@ -119,24 +183,52 @@ func ForwardAuth(c *gin.Context) {
...
+		return
+	}
+	if duration >= 0 {
+		if crowdsecBouncerCacheMode == "live" && duration != 0 {
+			cache.Set(key, []byte("t"), duration)
+		}
 		c.String(crowdsecBanResponseCode, crowdsecBanResponseMsg)
 	} else {
+		if crowdsecBouncerCacheMode == "live" {
+			cache.Set(key, []byte("f"), int(crowdsecBouncerDefaultCacheDuration.Seconds()))
```
[the remark this answers is not in the document]
Comment (maxlerebourg) [reply controler/controler.go]: Crowdsec can response with duration value when the IP is authorized ?

**Comment by maxlerebourg** [comment]
Hey, thanks for choosing my MR, the other one is made by a friend. We have a different approach to it, so you can go look at it too.
I have fixed all the issues you brought up, but one of them is not understandable to me.

Another point about the 403 return in case of the cache stream update fail. In your previous comment, you said "crash the bouncer" and in you PR comment "return 403 on /healthz", I choose the second, feel free to tell me the opposite was the good answer.

**Pull request closed without merging**